From 4d5bdf4b4a101f7afc4b3d52bce4395f412d5f06 Mon Sep 17 00:00:00 2001
From: Robin Gong <b38343@freescale.com>
Date: Mon, 8 Dec 2014 17:30:40 +0800
Subject: [PATCH 0183/5242] MLK-9768: dma: imx-sdma: fix UART loopback random
 failed

commit  ca92c6cd09b8798f491f14f32845a6a0db3a731c from
https://source.codeaurora.org/external/imx/linux-imx.git

For UART, we need use old chn_real_count to know the real rx count even in
cylic dma mode, because UART driver use cyclic mode to increase performance
without any data loss.

Signed-off-by: Robin Gong <b38343@freescale.com>
(cherry picked from commit 398cee2ad110c4f183e553af0564fbdcbe8548cb)
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/dma/imx-sdma.c |    7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/drivers/dma/imx-sdma.c b/drivers/dma/imx-sdma.c
index 09cf3d4..538f4bb 100644
--- a/drivers/dma/imx-sdma.c
+++ b/drivers/dma/imx-sdma.c
@@ -1435,7 +1435,12 @@ static enum dma_status sdma_tx_status(struct dma_chan *chan,
 	struct sdma_channel *sdmac = to_sdma_chan(chan);
 	u32 residue;
 
-	if (sdmac->flags & IMX_DMA_SG_LOOP)
+	/*
+	 * For uart rx data may not receive fully, use old chn_real_count to
+	 * know the real rx count.
+	 */
+	if ((sdmac->flags & IMX_DMA_SG_LOOP) &&
+	    sdmac->peripheral_type != IMX_DMATYPE_UART)
 		residue = (sdmac->num_bd - sdmac->buf_ptail) *
 			   sdmac->period_len - sdmac->chn_real_count;
 	else
-- 
1.7.9.5

